common/crypto/aes256.cc

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2024 Centreon
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ */
+#include <absl/container/fixed_array.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
+
+#include "aes256.hh"
+#include "base64.hh"
+#include "com/centreon/exceptions/msg_fmt.hh"
+
+namespace com::centreon::common::crypto {
+
+/**
+ * @brief The aes256 constructor. Th
+ *
+ * @param second_key
+ */
+aes256::aes256(const std::string& first_key, const std::string& second_key)
+    : _first_key{base64_decode(first_key)},
+      _second_key(base64_decode(second_key)) {
+  if (_first_key.size() != 32)
+    throw exceptions::msg_fmt(
+        "the key for aes256 must have a size of 256 bits and not {}",
+        _first_key.size() * 8);
+  assert(!_second_key.empty());
+}
+
+std::string aes256::encrypt(const std::string& input) {
+  const int iv_length = EVP_CIPHER_iv_length(EVP_aes_256_cbc());
+
+  uint32_t crypted_size =
+      input.size() + EVP_CIPHER_block_size(EVP_aes_256_cbc());
+
+  std::string result;
+  /* Here is the result just before the base64 encoding. It is composed of:
+   * * the iv
+   * * the hash of size 64
+   * * the crypted vector
+   * We reserve result to be suffisantly big to contain them. crypted_size
+   * is an estimated size, always bigger than the real result.
+   */
+  result.resize(iv_length + 64 + crypted_size);
+
+  std::string_view iv(result.data(), iv_length);
+  std::string_view hmac(result.data() + iv_length, 64);
+  std::string_view crypted_vector(result.data() + iv_length + 64, crypted_size);
+
+  RAND_bytes((unsigned char*)iv.data(), iv_length);
+
+  EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
```

Comment:
Is it expensive to create a ctx. If yes, why not store ctx in an attribute of this class?

Reply:
not really. And moreover, the ctx is not initialized the same way if we encrypt or decrypt so difficult to mutualize its declaration.
I saw some people doing it but that had to declare more things in the ctx, I'm not sure it's worth.